### PR TITLE
Geo-compose: add identify methods and microapp

### PR DESCRIPTION
**Pull request by hud10837**
-Adds all `identify` methods to the `GeoViewProxy`
    - These are added as methods with default parameters rather than the overload pattern used in view-based GeoView
    - `identifyLayer` and `identifyGraphicsOverlay` are both renamed to be `identify` overloads to match swift
- Modifies `GeoViewProxy` and `MapViewProxy` to ensure only one `GeoView` is cached and accessed
- Adds a microapp to demonstrate `identify` and use of the `MapViewProxy` by tapping on recent earthquake events.

### CODE REVIEW

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt [hunk 1]
```diff
@@ -17,10 +17,201 @@
...
+            screenCoordinate, tolerance, returnPopupsOnly, maximumResults
+        ) ?: Result.failure(IllegalStateException(nullGeoViewErrorMessage))
+    }
+
+    /**
+     * Initiate an Identify operation on the specified [layer].
+     *
+     * The [tolerance] parameter determines the extent of the region used during the identify operation. Any GeoElements
+     * that intersect this test region when rendered are returned, up to the [maximumResults] limit. A tolerance of
+     * 0 tests just the physical pixel at [screenCoordinate]. Tolerance values above 0 are in DIPs and specify a circular
```
Comment (gunt0001) [inline toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt:150]:
```suggestion
     * 0 tests just the physical pixel at [screenCoordinate]. Tolerance values above 0 are in [Dp] and specify a circular
```

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt [hunk 2]
```diff
@@ -17,10 +17,201 @@
...
+     * region centered on [screenCoordinate], with radius equal to [tolerance]. The maximum allowed tolerance value is 100 DIPs,
```
Comment (gunt0001) [inline toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt:151]:
```suggestion
     * region centered on [screenCoordinate], with radius equal to [tolerance]. The maximum allowed tolerance value is 100 Dp,
```

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt [hunk 3]
```diff
@@ -17,10 +17,201 @@
...
+     * resulting in an identify circle of diameter 200 DIPs. The [returnPopupsOnly] parameter controls what properties
```
Comment (gunt0001) [inline toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt:152]:
```suggestion
     * resulting in an identify circle of diameter 200 Dp. The [returnPopupsOnly] parameter controls what properties
```

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt [hunk 4]
```diff
@@ -17,10 +17,201 @@
...
+     * return per layer or sublayer. -1 indicates unlimited results
+     * @return A [Result] containing an [IdentifyLayerResult], or failure
+     * @since 200.4.0
+     */
+    public suspend fun identify(
+        layer: Layer,
+        screenCoordinate: ScreenCoordinate,
+        tolerance: Double,
+        returnPopupsOnly: Boolean = false,
+        maximumResults: Int = 1
```
Comment (gunt0001) [inline toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt:175]: As above, see [Swift's implementation](https://devtopia.esri.com/runtime/swift/blob/main/ArcGIS/Sources/ArcGIS/Views/GeoViewProxy.swift#L116).

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt [hunk 5]
```diff
@@ -17,10 +17,201 @@
...
+    }
+
+    /**
+     * Initiate an Identify operation on all layers in the view.
+     *
+     * Results are returned in top-to-bottom order.
+     *
+     * The [tolerance] parameter determines the extent of the region used during the identify operation. Any GeoElements
+     * that intersect this test region when rendered are returned, up to the [maximumResults] limit. A tolerance of 0 tests just the
+     * physical pixel at [screenCoordinate]. Tolerance values above 0 are in DIPs and specify a circular region centered on
```
Comment (gunt0001) [inline toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt:189]:
```suggestion
     * physical pixel at [screenCoordinate]. Tolerance values above 0 are in [Dp] and specify a circular region centered on
```

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt [hunk 6]
```diff
@@ -17,10 +17,201 @@
...
+     * [screenCoordinate], with radius equal to [tolerance]. The maximum allowed tolerance value is 100 DIPs, resulting in an
```
Comment (gunt0001) [inline toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt:190]:
```suggestion
     * [screenCoordinate], with radius equal to [tolerance]. The maximum allowed tolerance value is 100 Dp, resulting in an
```

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt [hunk 7]
```diff
@@ -17,10 +17,201 @@
...
+     * identify circle of diameter 200 DIPs. The [returnPopupsOnly] parameter controls what properties are populated in
```
Comment (gunt0001) [inline toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt:191]:
```suggestion
     * identify circle of diameter 200 Dp. The [returnPopupsOnly] parameter controls what properties are populated in
```

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt [hunk 8]
```diff
@@ -17,10 +17,201 @@
...
+     * @return A [Result] containing a [List] of [IdentifyLayerResult], containing one entry for each layer in the
+     * view that supports identify, or failure. Each entry contains a [Layer] and a [List] of elements of the type
+     * contained by the layer (e.g. [com.arcgismaps.data.Feature] for an [com.arcgismaps.mapping.layers.FeatureLayer])
+     * @since 200.4.0
+     */
+    public suspend fun identifyLayers(
+        screenCoordinate: ScreenCoordinate,
+        tolerance: Double,
+        returnPopupsOnly: Boolean = false,
+        maximumResults: Int = 1
```
Comment (gunt0001) [inline toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt:211]: As above, see [Swift's implementation](https://devtopia.esri.com/runtime/swift/blob/main/ArcGIS/Sources/ArcGIS/Views/GeoViewProxy.swift#L143).

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapViewProxy.kt [hunk 9]
```diff
@@ -35,23 +36,15 @@ import com.arcgismaps.mapping.view.ScreenCoordinate
...
-     * when it enters the composition and set to null when it is disposed.
-     *
-     * @since 200.3.0
-     */
-    internal fun setMapView(mapView: com.arcgismaps.mapping.view.MapView?) {
-        this.mapView = mapView
-    }
+    private val mapView: MapView?
+        get() {
+            return this.geoView as MapView?
```
Comment (gunt0001) [inline toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapViewProxy.kt:46]: This means you need to do a cast every time your reference the `mapView` property. I think what you had before with the `setMapView()` function was better. `setMapView()` can call an internal `setGeoView()` function on its parent class...

File: microapps/MapViewIdentifyApp/app/src/main/java/com/arcgismaps/toolkit/mapviewidentifyapp/screens/IdentifyViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+     *
+     * @since 200.4.0
+     */
+    val arcGISMap = ArcGISMap(BasemapStyle.ArcGISDarkGray).apply {
+        viewModelScope.launch(Dispatchers.Main) {
+            // Show a loading indicator while the feature layer loads
+            showProgressIndicator = true
+            featureLayer.load()
+            operationalLayers.add(featureLayer)
+            showProgressIndicator = false
```
[the remark this answers is not in the document]
Comment (hud10837) [reply microapps/MapViewIdentifyApp/app/src/main/java/com/arcgismaps/toolkit/mapviewidentifyapp/screens/IdentifyViewModel.kt]: I changed this by changing the message when no feature has been identified, and I removed the loading indicator here.

File: microapps/MapViewIdentifyApp/app/src/main/java/com/arcgismaps/toolkit/mapviewidentifyapp/screens/MainScreen.kt [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+import com.arcgismaps.toolkit.geocompose.MapView
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun MainScreen() {
+    val identifyViewModel: IdentifyViewModel = viewModel()
+    val bottomSheetScaffoldState = rememberBottomSheetScaffoldState()
+    // Expand the bottom sheet whenever the attributes to display are changed
+    LaunchedEffect(key1 = identifyViewModel.identifiedAttributes) {
+        bottomSheetScaffoldState.bottomSheetState.expand()
```
[the remark this answers is not in the document]
Comment (hud10837) [reply microapps/MapViewIdentifyApp/app/src/main/java/com/arcgismaps/toolkit/mapviewidentifyapp/screens/MainScreen.kt:60]: This is fixed by setting the height of the content. I also changed the initial viewpoint so the background grid isn't the only thing visible above the bottom sheet

File: microapps/MapViewIdentifyApp/app/src/main/java/com/arcgismaps/toolkit/mapviewidentifyapp/screens/MainScreen.kt [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+    }
```
Comment (gunt0001) [inline microapps/MapViewIdentifyApp/app/src/main/java/com/arcgismaps/toolkit/mapviewidentifyapp/screens/MainScreen.kt:61]: I think the use of bottom sheet might also be a good opportunity to showcase the use of insets, i.e. when the bottom sheet is expanded, adjust the MapView's bottom inset to match the height of the expanded bottom sheet. I tried make this work like this:
```kotlin
    val insets = remember { mutableStateOf(PaddingValues()) }
    val identifyViewModel: IdentifyViewModel = viewModel()
    val bottomSheetScaffoldState = rememberBottomSheetScaffoldState()
    // Expand the bottom sheet whenever the attributes to display are changed
    LaunchedEffect(key1 = identifyViewModel.identifiedAttributes) {
        bottomSheetScaffoldState.bottomSheetState.expand()
    }
    LaunchedEffect(bottomSheetScaffoldState.bottomSheetState.currentValue) {
        snapshotFlow { bottomSheetScaffoldState.bottomSheetState.currentValue }.collect { sheetValue ->
            when (sheetValue) {
                SheetValue.Expanded -> {
                    insets.value = PaddingValues(bottom = 350.dp)
                }
                else -> {
                    insets.value = PaddingValues()
                }
            }
        }
    }
```

File: microapps/MapViewIdentifyApp/app/src/main/java/com/arcgismaps/toolkit/mapviewidentifyapp/screens/MainScreen.kt [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+@Composable
+private fun IdentifyDetails(
+    showProgressIndicator: Boolean,
+    identifiedAttributes: Map<String, Any?>
+) {
+    Column(
+        Modifier
+            .fillMaxWidth()
+            // ensures the bottom sheet does not cover the screen
+            .height(350.dp),
```
Comment (gunt0001) [inline microapps/MapViewIdentifyApp/app/src/main/java/com/arcgismaps/toolkit/mapviewidentifyapp/screens/MainScreen.kt:105]: Is there an easy way to set this to 50% of screen height? I noticed that when you put the device in landscape mode, the bottom sheet would cover all or most of the MapView with the 350dp height.

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt [hunk 14]
```diff
@@ -17,10 +17,218 @@
...
+    public suspend fun identify(
+        graphicsOverlay: GraphicsOverlay,
+        screenCoordinate: ScreenCoordinate,
+        tolerance: Dp,
+        returnPopupsOnly: Boolean = false,
+        maximumResults: Int? = 1
+    ): Result<IdentifyGraphicsOverlayResult> {
+        val maximumResultsSanitized = if (maximumResults == null || maximumResults < 0) {
+            -1
+        } else maximumResults
```
Comment (gunt0001) [inline toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt:93]: Could this be refactored into a private helper method, something like `clampMaximumResults()` and then reuse on each function.

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt [hunk 15]
```diff
@@ -17,10 +17,218 @@
...
+     * @since 200.4.0
+     */
+    public suspend fun identify(
+        graphicsOverlay: GraphicsOverlay,
+        screenCoordinate: ScreenCoordinate,
+        tolerance: Dp,
+        returnPopupsOnly: Boolean = false,
+        maximumResults: Int? = 1
+    ): Result<IdentifyGraphicsOverlayResult> {
+        val maximumResultsSanitized = if (maximumResults == null || maximumResults < 0) {
```
Comment (gunt0001) [inline toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt:91]: I think it should be:
```suggestion
        val maximumResultsSanitized = if (maximumResults == null || maximumResults <= 0) {
```

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt [hunk 16]
```diff
@@ -17,10 +17,218 @@
...
+        ) ?: Result.failure(IllegalStateException(nullGeoViewErrorMessage))
+    }
+
+    /**
+     * Initiate an Identify operation on the specified [layer].
+     *
+     * The [tolerance] parameter determines the extent of the region used during the identify operation. Any GeoElements
+     * that intersect this test region when rendered are returned, up to the [maximumResults] limit. A tolerance of
+     * 0 tests just the physical pixel at [screenCoordinate]. Tolerance values above 0 are in [Dp] and specify a circular
+     * region centered on [screenCoordinate], with radius equal to [tolerance]. The maximum allowed tolerance value is 100 [Dp],
```
Comment (gunt0001) [inline toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt:153]: Linking to [Dp] once in the doc is probably enough:
```suggestion
     * region centered on [screenCoordinate], with radius equal to [tolerance]. The maximum allowed tolerance value is 100 Dp,
```

Would need to be updated in other places as well.

**Review by gunt0001 — COMMENTED** [review]
@hud10837 - looks great, a few more comments/suggestions.

File: microapps/MapViewIdentifyApp/app/src/main/java/com/arcgismaps/toolkit/mapviewidentifyapp/screens/IdentifyViewModel.kt [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        currentIdentifyJob?.cancel()
+        currentIdentifyJob = viewModelScope.launch {
+            showProgressIndicator = true
+            identifiedAttributes = emptyMap()
+            featureLayer.clearSelection()
+            val result =
+                mapViewProxy.identify(featureLayer, singleTapConfirmedEvent.screenCoordinate, 20.dp)
+            // yielding here will ensure that if the job gets cancelled, we don't try to set the
+            // identified attributes
+            yield()
```
Comment (gunt0001) [inline microapps/MapViewIdentifyApp/app/src/main/java/com/arcgismaps/toolkit/mapviewidentifyapp/screens/IdentifyViewModel.kt:86]: Is this necessary? If the job is cancelled, the result returned from `identify()` should be a failure anyway.

File: microapps/MapViewIdentifyApp/app/src/main/java/com/arcgismaps/toolkit/mapviewidentifyapp/screens/MainScreen.kt [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+@Composable
+private fun IdentifyDetails(
+    showProgressIndicator: Boolean,
+    identifiedAttributes: Map<String, Any?>
+) {
+    Column(
+        Modifier
+            .fillMaxWidth()
+            // ensures the bottom sheet doesn't cover the whole screen
+            .heightIn(max = 200.dp),
```
Comment (gunt0001) [inline microapps/MapViewIdentifyApp/app/src/main/java/com/arcgismaps/toolkit/mapviewidentifyapp/screens/MainScreen.kt:128]: Could you hold this 200dp in a const and re-use that in the snapshotFlow above.

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt [hunk 19]
```diff
@@ -17,10 +17,212 @@
...
+import com.arcgismaps.mapping.view.IdentifyLayerResult
+import com.arcgismaps.mapping.view.ScreenCoordinate
+
 /**
  * Used to perform operations on a composable MapView or SceneView.
  *
  * @since 200.3.0
  */
-public sealed class GeoViewProxy {
+public sealed class GeoViewProxy(private val classNameForErrorMessage: String) {
```
Comment (gunt0001) [inline toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GeoViewProxy.kt:33]: naming suggestion:
```suggestion
public sealed class GeoViewProxy(private val className: String) {
```

**Review by gunt0001 — APPROVED** [review]
@hud10837 - looks great. A few minor comments.

**Review by puneet-pdx — COMMENTED** [review]
@hudson I am seeing an issue where on minimizing the bottomsheet the attribution bar does not reposition and stays in the same position once you tap on a feature to identify it. 
 

https://github.com/Esri/arcgis-maps-sdk-kotlin-toolkit/assets/5582469/a41992ea-4fb1-4e2d-967d-f798ac38b59c

**Comment by hud10837** [comment]
@puneet-pdx The insets behavior was overcomplicating this microapp, so I have removed it entirely and it can potentially be covered as a follow-up

**Review by puneet-pdx — APPROVED** [review]
@hud10837 you missed to address Gunther's comments that he left when he approved the PR. I am approving this so you can address Gunthers comments and merge it.

**Comment by hud10837** [comment]
Thanks @gunt0001 @puneet-pdx I have made all those changes

**Pull request merged**